### DISCUSSION TITLE
[MueLu] Working example of xml for exporting aggregates when using BlockedRAPFactory?

### DISCUSSION THREAD

**Discussion by nasseralkmim** [trilinos/Trilinos#12143] [Q&A] [answered]
Hi,

I have a working example of exporting aggregates with a regular non-block matrix.
However when dealing with block systems (multi-field problems) it is not clear if this feature is available. 
I tried to pass the AggregationExportFactory as part of the TransferFactories in the BlockedRAPFactory, but it does not work.

With the block system, it is also not clear how one should pass the coordinates to the hierarchy level 0. I tried with the "user data" in the parameter list and with the hierarchy object itself.

Does anyone have experience with this?

**Comment by nasseralkmim** [comment, marked as answer]
I manage to get this working. Here are some note about it.
The AggregationExportFactory<sup><a id="fnr.1" class="footref" href="#fn.1" role="doc-backlink">1</a></sup> requires the Coordinates on level 0.

```
  what():  AggregationExportFactory::DeclareInput: (LevelID = 0) unable to find or generate requested data "Coordinates" with generating factory "null" [0]
    during request for data "    Coordinates" on level 0 by factory  AggregationExportFactory [0x5627c311bf40]
    during request for data "              A" on level 1 by factory                 NoFactory [0x5627c338a440]
```

The Coordinates can be passed to the hierarchy object.

```C++
hierarchy->GetLevel(0)->Set("Coordinates", xcoordinates);
```

Then, in the tentative prolongation operator construction, it depends on the null space of the finer level<sup><a id="fnr.2" class="footref" href="#fn.2" role="doc-backlink">2</a></sup>. The null space factory can request coordinates if `nullspace: calculate rotations` the option is set to true.

```
  what():  NullspaceFactory::DeclareInput: (LevelID = 0) unable to find or generate requested data "Coordinates" with generating factory "null" [0]
    during request for data "    Coordinates" on level 0 by factory          NullspaceFactory [0x55560d90d270]
    during request for data "      Nullspace" on level 0 by factory         TentativePFactory [0x55560d93ef90]
    during request for data "              P" on level 1 by factory           BlockedPFactory [0x55560d682670]
    during request for data "              P" on level 1 by factory                 NoFactory [0x55560d93e9f0]
```

Without this option, the tentative prolongation factory itself can request for coordinates if the option to generate coarse levels coordinates is true `tentative: build coarse coordinates` and the coordinates is available in the finer level<sup><a id="fnr.3" class="footref" href="#fn.3" role="doc-backlink">3</a></sup>. This option seems to be true by default<sup><a id="fnr.4" class="footref" href="#fn.4" role="doc-backlink">4</a></sup>.

```
  what():  TentativePFactory::DeclareInput: (LevelID = 0) unable to find or generate requested data "Coordinates" with generating factory "null" [0]
    during request for data "    Coordinates" on level 0 by factory         TentativePFactory [0x55a19e28b1a0]
    during request for data "              P" on level 1 by factory           BlockedPFactory [0x55a19dfcf670]
    during request for data "              P" on level 1 by factory                 NoFactory [0x55a19e2694f0]
```

In this case, the tentative prolongation factory for a block is looking for the `Coordinates` variable. This variable can be informed to the specific block with the `FineLevelInputDataFactory` and declaring it in the group manager.

```xml
<ParameterList name="myInputCoordsFact">
    <Parameter name="factory" type="string" value="FineLevelInputDataFactory"/>
    <Parameter name="Variable" type="string" value="Coordinates"/>
    <Parameter name="Variable type" type="string" value="MultiVector"/>
    <Parameter name="Fine level factory" type="string" value="NoFactory"/>
</ParameterList>

<ParameterList name="myFirstGroup">
    <Parameter name="group" type="string" value="FactoryManager"/>
    <Parameter name="A" type="string" value="mySubBlockAFactory1"/>
    <Parameter name="P" type="string" value="myTentativePFact1"/>
    <Parameter name="R" type="string" value="myRFact1"/>
    <Parameter name="Aggregates" type="string" value="myAggFact1"/>
    <Parameter name="Nullspace" type="string" value="myNspFact1"/>
    <Parameter name="CoarseMap" type="string" value="myCoarseMap1"/>
    <Parameter name="Coordinates" type="string" value="myInputCoordsFact"/>
</ParameterList>
```

The aggregate export factory needs to declared as a Transfer factory in the blocked RAP factory,

```xml
<ParameterList name="myBlockedRAPFact">
    <!-- Just use the blocked prolongation and restriction -->
    <Parameter name="factory" type="string" value="BlockedRAPFactory"/>
    <Parameter name="P" type="string" value="myBlockedPFact"/>
    <Parameter name="R" type="string" value="myBlockedRFact"/>
    <!-- RAP factory is responsible to CREATE THE COARSER level matrices. -->
    <!-- It contains "TransferFactories" that transfer information from fine to coarse levels. -->
    <ParameterList name="TransferFactories">
    <Parameter name="For Coordinates" type="string" value="myTransferCoordinatesFact"/>
    <Parameter name="AggregationExportFactory"          type="string" value="myAggExport"/>
    </ParameterList>
</ParameterList>
```

## Footnotes

<sup><a id="fn.1" class="footnum" href="#fnr.1">1</a></sup> [file:~/.local/src/Trilinos/packages/muelu/src/Utils/MueLu<sub>AggregationExportFactory</sub><sub>def.hpp</sub>:://Only pull in coordinates if the user explicitly requests direct VTK output, so as not to break uses of old code](file:///home/nasser/.local/src/Trilinos/packages/muelu/src/Utils/MueLu_AggregationExportFactory_def.hpp)

<sup><a id="fn.2" class="footnum" href="#fnr.2">2</a></sup> [file:~/.local/src/Trilinos/packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu<sub>TentativePFactory</sub><sub>def.hpp</sub>::pName;](file:///home/nasser/.local/src/Trilinos/packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_TentativePFactory_def.hpp)

<sup><a id="fn.3" class="footnum" href="#fnr.3">3</a></sup> [file:~/.local/src/Trilinos/packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu<sub>TentativePFactory</sub><sub>def.hpp</sub>::if](file:///home/nasser/.local/src/Trilinos/packages/muelu/src/Transfers/Smoothed-Aggregation/MueLu_TentativePFactory_def.hpp)

<sup><a id="fn.4" class="footnum" href="#fnr.4">4</a></sup> [file:~/.local/src/Trilinos/packages/muelu/src/MueCentral/MueLu<sub>MasterList.cpp</sub>::"<Parameter name=\\"tentative: build coarse coordinates\\" type=\\"bool\\" value=\\"true\\"/>"](file:///home/nasser/.local/src/Trilinos/packages/muelu/src/MueCentral/MueLu_MasterList.cpp)

**Comment by cgcgcg** [comment]
@nasseralkmim Thank you for posting notes on how you got it to work. We're currently all on travel and unable to look into it.
